### PR TITLE
Show individual process' memory usage (RSS)

### PR DESCRIPTION
**Pull request by pmuetschard**
If the RSS (resident set size) counters are enabled, the kernel emits ftrace events when pages are added/removed from the various memory pools. A resident page is a page that is occupying real physical memory and so RSS is a good indicator of an app's memory usage. This changes adds a memory track similar to the system memory summary track to each process where we have RSS counter data showing the process' memory usage.

Bug: http://b/143885588

### CODE REVIEW

File: gapic/src/main/com/google/gapid/perfetto/models/BatterySummaryTrack.java [hunk 1]
```diff
@@ -15,135 +15,40 @@
...
+import java.util.function.BinaryOperator;
 
-  private final long capacityId;
-  private final long chargeId;
-  private final long currentId;
+public class BatterySummaryTrack
+    extends CombinedCountersTrack<BatterySummaryTrack.Data, BatterySummaryTrack.Values> {
   private final double maxAbsCurrent;
-  private final boolean needQuantize;
+  //private final boolean needQuantize;
```
[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapic/src/main/com/google/gapid/perfetto/models/BatterySummaryTrack.java]: done.

File: gapic/src/main/com/google/gapid/perfetto/models/CombinedCountersTrack.java [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    }
+
+    long[] ts = new long[a.ts.length + b.ts.length];
+    int count = combineTs(a.ts, b.ts, ts);
+    V v = createValues(count, this::combine);
+    System.arraycopy(ts, 0, v.ts, 0, count);
+
+    int ai = 0, bi = 0, ri = 0;
+    for (; ai < a.ts.length && bi < b.ts.length; ri++) {
+      long av = a.ts[ai], bv = b.ts[bi], rv = v.ts[ri];
```
[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapic/src/main/com/google/gapid/perfetto/models/CombinedCountersTrack.java]: done.

File: gapic/src/main/com/google/gapid/perfetto/models/CombinedCountersTrack.java [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+      }
+    }
+    sb.append(" from counter where track_id in ");
+    sb.append(stream(columns)
+        .filter(c -> !c.isNil())
+        .map(c -> Long.toString(c.trackId))
+        .collect(joining(", ", "(", ")")));
+    sb.append(" group by ts) window win as (order by ts)");
+    return sb.toString();
+  }
```
[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapic/src/main/com/google/gapid/perfetto/models/CombinedCountersTrack.java]: Done.

File: gapic/src/main/com/google/gapid/perfetto/views/ProcessMemoryPanel.java [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+      List<Integer> visibleSelected = Lists.newArrayList();
+      long maxUsage = track.getMaxUsage(), maxSwap = Math.max(maxUsage, track.getMaxSwap());
+
+      memoryRssSharedGradient().applyBase(ctx);
+      ctx.path(path -> {
+        path.moveTo(0, h);
+        double lastX = 0, lastY = h;
+        for (int i = 0; i < data.ts.length; i++) {
+          double nextX = state.timeToPx(data.ts[i]);
+          double nextY = h - (h * (data.file[i] + data.anon[i] + data.shared[i]) / maxUsage);
```
[the remark this answers is not in the document]
Comment (pmuetschard) [reply gapic/src/main/com/google/gapid/perfetto/views/ProcessMemoryPanel.java:97]: Swap is separate. By definition swapped memory is not resident. The UI shows two graphs: a stacked graph showing the resident memory used broken down by the three categories and a line graph showing the swap usage.

[the remark this answers is not in the document]
Comment (stellama0208) [reply gapic/src/main/com/google/gapid/perfetto/views/ProcessMemoryPanel.java:97]: I see, thanks for explaining.

**Pull request merged**